hudi-client/src/main/java/org/apache/hudi/index/HoodieGlobalSimpleIndex.java

```diff
@@ -0,0 +1,157 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.index;
+
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.common.model.EmptyHoodieRecordPayload;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.model.HoodieRecordLocation;
+import org.apache.hudi.common.model.HoodieRecordPayload;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.config.HoodieWriteConfig;
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.hudi.table.HoodieTable;
+
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import scala.Tuple2;
+
+import static java.util.stream.Collectors.toList;
+import static org.apache.hudi.index.HoodieIndexUtils.loadLatestDataFilesForAllPartitions;
+
+/**
+ * A global simple index which reads interested fields(record key and partition path) from base files and
+ * joins with incoming records to find the tagged location.
+ *
+ * @param <T>
+ */
+public class HoodieGlobalSimpleIndex<T extends HoodieRecordPayload> extends HoodieSimpleIndex<T> {
+
+  public HoodieGlobalSimpleIndex(HoodieWriteConfig config) {
+    super(config);
+  }
+
+  @Override
+  public JavaRDD<HoodieRecord<T>> tagLocation(JavaRDD<HoodieRecord<T>> recordRDD, JavaSparkContext jsc,
+                                              HoodieTable<T> hoodieTable) {
+    return tagLocationInternal(recordRDD, jsc, hoodieTable);
+  }
+
+  /**
+   * Tags records location for incoming records.
+   *
+   * @param recordRDD   {@link JavaRDD} of incoming records
+   * @param jsc         instance of {@link JavaSparkContext} to use
+   * @param hoodieTable instance of {@link HoodieTable} to use
+   * @return {@link JavaRDD} of records with record locations set
+   */
+  protected JavaRDD<HoodieRecord<T>> tagLocationInternal(JavaRDD<HoodieRecord<T>> recordRDD, JavaSparkContext jsc,
+                                                         HoodieTable<T> hoodieTable) {
+    JavaPairRDD<HoodieKey, HoodieRecord> incomingRecords = recordRDD.mapToPair(entry -> new Tuple2<>(entry.getKey(), entry));
+
+    JavaPairRDD<HoodieKey, HoodieRecordLocation> existingRecords = fetchRecordLocations(incomingRecords.keys(), jsc, hoodieTable,
```

Comment:
@nsivabalan This is problematic.. since we don't cache (rightfully so) the input here for Global index, 

```
JavaPairRDD<String, String> partitionRecordKeyPairRDD = hoodieKeys.mapToPair(entry -> new Tuple2(entry.getPartitionPath(), entry.getRecordKey()));
    List<String> affectedPartitionPathList = partitionRecordKeyPairRDD.map(tuple -> tuple._1).distinct().collect();
```

will collect() once reading the input, and then later ` return getTaggedRecords(incomingRecordsKeyedOnRecordKeys, existingRecordsKeyedOnRecordKeys);` will also read the input again for joining.. 

Also why `fetchRecordLocations()` we should be loading all the partitions, not just the ones based on incomingRecord, right? 

Global should be much simpler.. All it needs is to create `existingRecords` based on listing from all partitions and then do the getTaggedRecords() join..

Reply:
I have tried to do a bit of simplifying. But wanted to remind you that, we need to have recordKey as key to join and not HoodieKey since partition path may diff for existing keys compared to incoming records and hence. And so, from the return value of fetchRecordLocations, I had to do another mapToPair.

Reply:
Wrt fetchRecordLocations(), we have overloaded loadAllFilesForPartitions() in HoodieGlobalSimpleIndex. So, we are loading all partitions.